helper_crates/vtable/src/vrc.rs

```diff
@@ -102,25 +101,27 @@ pub struct VRc<VTable: VTableMetaDropInPlace + 'static, X = Dyn> {
 impl<VTable: VTableMetaDropInPlace + 'static, X> Drop for VRc<VTable, X> {
     fn drop(&mut self) {
         unsafe {
-            let inner = self.inner.as_ref();
-            if inner.strong_ref.fetch_sub(1, Ordering::SeqCst) == 1 {
-                let data = inner.data_ptr() as *const _ as *const u8 as *mut u8;
-                let mut layout = VTable::drop_in_place(inner.vtable, data);
+            let inner = self.inner.as_ptr();
+            if (*inner).strong_ref.fetch_sub(1, Ordering::SeqCst) == 1 {
+                let data =
+                    (inner as *mut u8).add(*core::ptr::addr_of!((*inner).data_offset) as usize);
+                let vtable = core::ptr::addr_of!((*inner).vtable);
+                let mut layout = VTable::drop_in_place(&*vtable, data);
                 layout = core::alloc::Layout::new::<VRcInner<VTable, ()>>()
                     .extend(layout.try_into().unwrap())
                     .unwrap()
                     .0
                     .pad_to_align()
                     .into();
-                if inner.weak_ref.load(Ordering::SeqCst) > 1 {
+                if (*core::ptr::addr_of!((*inner).weak_ref)).load(Ordering::SeqCst) > 1 {
                     // at this point we are sure that no other thread can access the data
                     // since we still hold a weak reference, so the other weak references
                     // in other thread won't start destroying the object.
-                    self.inner.cast::<VRcInner<VTable, Layout>>().as_mut().data = layout;
+                    *(VRcInner::data_ptr(self.inner.cast::<VRcInner<VTable, Layout>>().as_ptr())
+                        as *mut Layout) = layout;
                 }
-                if inner.weak_ref.fetch_sub(1, Ordering::SeqCst) == 1 {
```

Comment:
I’m curious: what’s the intermediate reference that’s created where the types don’t match?